src/android/com/eclipsesource/firebase/reference/Reference.java

```diff
@@ -0,0 +1,182 @@
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.support.v4.content.LocalBroadcastManager;
+import android.util.Log;
+
+import com.eclipsesource.tabris.android.RemoteObject;
+import com.eclipsesource.tabris.android.TabrisActivity;
+import com.eclipsesource.tabris.android.TabrisContext;
+import com.eclipsesource.tabris.android.internal.toolkit.AppState;
+import com.eclipsesource.tabris.android.internal.toolkit.IAppStateListener;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.google.firebase.database.ChildEventListener;
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.DatabaseError;
+import com.google.firebase.database.DatabaseReference;
+import com.google.firebase.database.FirebaseDatabase;
+import com.google.firebase.database.ValueEventListener;
+
+
+class Reference { //Represents DatabaseReference... or should it represent FirebaseDatabase ?
+    private DatabaseReference database;
+
+    private final Activity activity;
+    private final TabrisContext tabrisContext;
+    private String path;
+
+    Reference(Activity activity, TabrisContext context){
+        this.activity = activity;
+        this.tabrisContext = context;
+        Log.e("Nery", "Reference created");
+        if(this.activity == null){
+            Log.e("ERROR", "activity es null");
+        }
+        if(this.tabrisContext == null){
+            Log.e("ERROR", "tabrisContext es null");
+        }
+    }
+
+
+    public void create(DatabaseReference ref){
+        this.path = ref.toString();
+        this.database = ref;
+        registerListeners();
+    }
+
+
+    public Reference child(String path){
+        Reference ref = new Reference(this.activity, this.tabrisContext);
+        ref.create(this.database.child(path));
+        return ref;
+    }
+    
+    public Reference push(){
+        Reference ref = new Reference(this.activity, this.tabrisContext);
+        ref.create(this.database.push());
+        return ref;
+    }
+
+    public void keepSynced(boolean keepSynced){
+        this.database.keepSynced(keepSynced);
+    }
+
+    public void setValue(Object value){
+        //Set the value and add a CompletitionListener
+        database.setValue(value, completition);
+    }
+
+    public String getKey(){
+        return this.database.getKey();
+    }
+
+
+
+
+    private void registerListeners(){
+        database.addValueEventListener(valueListener);   
+
+        database.addChildEventListener(childlistener);
+    }
+    
+    private DatabaseReference.CompletionListener completition = new DatabaseReference.CompletionListener() {
+        @Override
+        public void onComplete(DatabaseError error, DatabaseReference ref){
```

Comment:
on the js side we could use a Promise to handle the result of a `CompletionListener`.